subsys/bluetooth/audio/has.c

```diff
@@ -0,0 +1,1330 @@
+			preset_changed_clear_all(client);
+		} else if (is_preset_changed_pending(client) &&
+			   atomic_test_bit(client->flags, CLIENT_FLAG_ATT_MTU_VALID)) {
+			control_point_tx_work_submit(client, BT_HAS_CP_WORK_TIMEOUT);
+		}
+	}
+}
+
+BT_CONN_CB_DEFINE(has_conn_callbacks) = {
+	.connected = connected,
+	.disconnected = disconnected,
+	.security_changed = security_changed,
+};
+
+static void att_mtu_updated(struct bt_conn *conn, uint16_t tx, uint16_t rx)
+{
+	struct has_client *client;
+
+	BT_DBG("conn %p tx %u rx %u", conn, tx, rx);
+
+	client = client_find(conn);
+	if (!client) {
+		return;
+	}
+
+	/* HearingAidProfile_validation_r01
+	 *
+	 * If the HARC supports the Read All Presets procedure or the Read Preset by Index
+	 * procedure, then the HARC shall support an ATT_MTU value no less than 49.
+	 */
+	if (tx < BT_HAS_ATT_MTU_MIN) {
+		return;
+	}
+
+	/* Send pending Preset Changed if any. */
+	if (!atomic_test_and_set_bit(client->flags, CLIENT_FLAG_ATT_MTU_VALID) &&
+	    is_preset_changed_pending(client) &&
+	    atomic_test_bit(client->flags, CLIENT_FLAG_ENCRYPTED) &&
+	    atomic_test_bit(client->flags, CLIENT_FLAG_CP_IND_ENABLED)) {
+		control_point_tx_work_submit(client, BT_HAS_CP_WORK_TIMEOUT);
+	}
+}
+
+static struct bt_gatt_cb gatt_cb = {
+	.att_mtu_updated = att_mtu_updated,
+};
+
+static int has_init(const struct device *dev)
+{
+	uint8_t type;
+
+	ARG_UNUSED(dev);
+
+	BUILD_ASSERT(BT_HAS_CP_TX_MAX_LEN < CONFIG_BT_L2CAP_TX_MTU);
+
+	if (IS_ENABLED(CONFIG_BT_HAS_HA_TYPE_MONAURAL)) {
+		type = BT_HAS_MONAURAL_HEARING_AID;
+	} else if (IS_ENABLED(CONFIG_BT_HAS_HA_TYPE_BANDED)) {
+		type = BT_HAS_BANDED_HEARING_AID;
+	} else {
+		type = BT_HAS_BINAURAL_HEARING_AID;
+	}
+
+	/* Initialize the supported features characteristic value */
+	has_local.features = BT_HAS_FEAT_HEARING_AID_TYPE_MASK & type;
+	if (IS_ENABLED(CONFIG_BT_HAS_PRESET_SYNC_SUPPORT)) {
+		has_local.features |= BT_HAS_FEAT_BIT_PRESET_SYNC;
+	}
+	if (!IS_ENABLED(CONFIG_BT_HAS_IDENTICAL_PRESET_RECORDS)) {
+		has_local.features |= BT_HAS_FEAT_BIT_INDEPENDENT_PRESETS;
+	}
+
+	bt_gatt_cb_register(&gatt_cb);
+
+	k_work_init(&active_preset_work, active_preset_work_process);
+
+	return 0;
+}
+
+SYS_INIT(has_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
+
+#define BT_HAS_IS_LOCAL(_has) ((_has) == &has_local)
+#endif /* CONFIG_BT_HAS */
+
+int bt_has_register(struct bt_has_register_param *param, struct bt_has **out)
+{
+#if defined(CONFIG_BT_HAS)
+	bool writable_presets_support = false;
+
+	*out = &has_local;
+
+	CHECKIF(param == NULL || param->ops == NULL || param->ops->active_set == NULL)
+	{
+		return -EINVAL;
+	}
+
+	/* Check if already registered */
+	if (preset_ops) {
+		return -EALREADY;
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(has_preset_list); i++) {
+		struct bt_has_preset_register_param *preset_param = NULL;
+		struct has_preset *preset = &has_preset_list[i];
+
+		/* Sort the presets in order of increasing ID */
```

Comment:
Is this required? Do we gain anything from doing that? 

```suggestion
		/* Sort the presets in order of increasing index */
```

Reply:
Yes, as the presets shall be sent to client in increasing order. So it's better to sort them on init